docs/design/013-caching/README.md

```diff
@@ -0,0 +1,103 @@
+# Caching of data from ledger
+
+Currently whenever credential definitions and/or schemas is needed, it is being fetched from the ledger.
+This operation may last multiple seconds and is slowing down usage of credentials.
+Caching also enables usage of anoncreds in areas where user do not have internet coverage (eg. Using passport credential on foreign airport).
+
+## Goals and ideas
+
+* Allow users to cache credential definitions and schemas.
+  * Local wallet to be used because although this data is public, possession of some credential definition or schema reveals possession of respective credential. 
+* Provide higher level api for fetching this data so it is easier to use.
+  * Caching should be transparent to the user.
+* Enable purging of old (not needed more) data.
+  
+## Public API
+
+Note: In all calls `pool_handle` may be removed if did resolver is implemented.
+
+```Rust
+/// Gets credential definition json data for specified credential definition id.
+/// If data is present inside of cache, cached data is returned.
+/// Otherwise data is fetched from the ledger and stored inside of cache for future use.
+/// 
+/// #Params
+/// command_handle: command handle to map callback to caller context.
+/// pool_handle: pool handle (created by open_pool_ledger).
+/// wallet_handle: wallet handle (created by open_wallet).
+/// submitter_did: DID of the submitter stored in secured Wallet.
+/// id: identifier of credential definition.
+/// cb: Callback that takes command result as parameter.
+#[no_mangle]
+pub extern fn indy_get_cred_def(command_handle: IndyHandle,
+                                pool_handle: IndyHandle,
+                                wallet_handle: IndyHandle,
+                                submitter_did: *const c_char,
```

Comment:
1. I suggest to remove submitter_did param as it can be hard-coded or auto-generated for each request
2. Do we need to return cache timestamp also?
3. I suggest to add options_json that will allow to:

- Skip caching
- Provide required freshness

Reply:
3. Skip caching have sense in a meaning do not store retrieved value inside of cache.
    For freshness, I think that schema and credential_def cannot be updated (without change of id) so freshness is not needed. For that reason I do not see need for cache timestamp.

Reply:
In my understanding CredDef and Schema can be updated. Schema can get backward compatible changes.  Keys in CredDef can be rotated in some cases. Also some entities we plan to cache like DID Doc can be updated much often. @ashcherbakov What do you think?